src/qibo/tomography/gate_set_tomography.py

```diff
@@ -0,0 +1,302 @@
+from functools import cache
+from inspect import signature
+from itertools import product
+from random import Random
+from typing import Union
+
+import numpy as np
+from networkx import connectivity
+from sympy import S
+
+from qibo import Circuit, gates, symbols
+from qibo.backends import GlobalBackend, _check_backend
+from qibo.config import raise_error
+from qibo.hamiltonians import SymbolicHamiltonian
+from qibo.transpiler.optimizer import Preprocessing
+from qibo.transpiler.pipeline import Passes
+from qibo.transpiler.placer import Random
+from qibo.transpiler.router import Sabre
+from qibo.transpiler.unroller import NativeGates, Unroller
+
+SUPPORTED_NQUBITS = [1, 2]
+"""Supported nqubits for GST."""
+
+
+def _check_nqubits(nqubits):
+    if nqubits not in SUPPORTED_NQUBITS:
+        raise_error(
+            ValueError,
+            f"nqubits given as {nqubits}. nqubits needs to be either 1 or 2.",
+        )
+
+
+@cache
+def _gates(nqubits):
+    """Gates implementing all the GST state preparations.
+
+    Args:
+        nqubits (int): Number of qubits for the circuit.
+    Returns:
+        list(:class:`qibo.gates.abstract.Gate`): list of the gates used to prepare the possible states.
+    """
+
+    return list(
+        product(
+            [(gates.I,), (gates.X,), (gates.H,), (gates.H, gates.S)], repeat=nqubits
+        )
+    )
+
+
+@cache
+def _measurements(nqubits):
+    """Measurement gates implementing all the GST measurement bases.
+
+    Args:
+        nqubits (int): Number of qubits for the circuit.
+    Returns:
+        list(:class:`qibo.gates.abstract.Gate`): list of the gates implementing the possible measurement bases.
+    """
+
+    return list(product([gates.Z, gates.X, gates.Y, gates.Z], repeat=nqubits))
+
+
+@cache
+def _observables(nqubits):
+    """All the observables measured in GST.
+
+    Args:
+        nqubits (int): Number of qubits for the circuit.
+    Returns:
+        list(:class:`qibo.symbols.Symbol`): list of the possible observables measured.
+    """
+
+    return list(product([symbols.I, symbols.Z, symbols.Z, symbols.Z], repeat=nqubits))
+
+
+@cache
+def _get_observable(j, nqubits):
+    """Returns the j-th observable. Here,
+
+    .. math::
+        j \\in \\{0, 1, 2, 3\\}^{\\otimes n} \\equiv \\{ I, X, Y, Z\\}^{\\otimes n}.
```

Comment:
The first set doesn't really make mathematical sense. Maybe you meant something like the suggestion below
```suggestion
    .. math::
        j \\in \\{"00", "01", "10", "11"\\}^{\\otimes n} \\equiv \\{ I, X, Y, Z\\}^{\\otimes n}.
```

Reply:
I am not sure about this, probably @mho291 is the best person to ask to

Reply:
j are labels. So j \in {0, 1, 2, 3} refers to measuring in the {I, X, Y, Z} bases. If we do the \otimes 2 for 2 qubit system, then we have j \in {0, 1, 2, 3, 4, ..., 15} refers to measuring in the {II, IX, IY, IZ, XI, ..., ZZ} bases.

Reply:
I'm aware of what it is supposed to mean. But it only means what's supposed to mean if you explicitly mention that :math:`j` is being represented in a base-:math:`4` string. This is not common practice, but it works. My advice is to either change it to the commonly used base-:math:`2` representation or to explicitly mention that it is base-:math:`4` indexing. Former is preferred but the latter works.

Reply:
I've not seen anyone use base 2 to index j and k but if you feel that base 2 is the way to go, then sure please go ahead and change it. It doesn't matter to me as long as the code works. Feel free to modify any mathematical notations which I inaccurately notate too.

Reply:
https://qibo.science/qibo/stable/api-reference/qibo.html#phase-space-representation-of-stabilizer-states

Reply:
Apologies if my reply sounded dismissive. Thanks for the link. I will take a look.

I encode the state preparation and measurement bases in base 4, or what seems to be base 4, because it is trivial. j \in {0, 1, 2, 3}^\otimes n refers to {I, X, Y, Z}^\otimes n. Now, when n=1, it's simple. When n=2, we're to find the expectation of {II, IX, IY, IZ, XI, ..., ZZ}. I need to correct myself in that j doesn't run from 0 to 15 but in fact {00, 01, 02, 03, ..., 33}, which makes more sense.

For k, the same applies. k \in {0, 1, 2, 3}^\otimes n refers to preparing these states {|0>, |1>, |+>, |y+>}^\otimes n. To be more precise, the preparing rho_k means preparing {|0><0|, |1><1|, |+><+|, |y+><y+|}^\otimes n. The same logic applies when we have n=2. We prepare {|00><00|, ..., |y+ y+><y+ y+|} which are indexed by {00, 01, 02, ..., 33}.

If we change it to base 2 notation, i.e. j \in {00, 01, 10, 11}, I find that it is less straightforward especially if we're looking at n=2 and higher (i.e. j \in {0000, 0001, 0010, 0011, ...., 1111} for n=2 case).

In any case, it's a trivial issue. If you think that base 2 is the way to go to be in line with the other modules in Qibo, then I will change it.

Reply:
Just one line after the equation explaining that you're representing `j` as a base-4 string suffices.

Reply:
Sure, I will do just that. Thanks Renato.